engine/src/main/java/io/camunda/zeebe/streamprocessor/records/RecordBatch.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.streamprocessor.records;
+
+import io.camunda.zeebe.protocol.record.RecordType;
+import io.camunda.zeebe.protocol.record.RejectionType;
+import io.camunda.zeebe.protocol.record.ValueType;
+import io.camunda.zeebe.protocol.record.intent.Intent;
+import io.camunda.zeebe.util.buffer.BufferWriter;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+
+public final class RecordBatch implements ModifiableRecordBatch {
+
+  final List<UnmodifiableRecordBatchEntry> recordBatchEntries = new ArrayList<>();
+  private int batchSize;
+  private final int maxBatchSize;
+
+  public RecordBatch(final int maxBatchSize) {
+    this.maxBatchSize = maxBatchSize;
+  }
+
+  @Override
+  public void appendRecord(
+      final long key,
+      final int sourceIndex,
+      final RecordType recordType,
+      final Intent intent,
+      final RejectionType rejectionType,
+      final String rejectionReason,
+      final ValueType valueType,
+      final BufferWriter valueWriter) {
+    final var recordBatchEntry =
+        RecordBatchEntry.createRecordBatchEntry(
+            key,
+            sourceIndex,
+            recordType,
+            intent,
+            rejectionType,
+            rejectionReason,
+            valueType,
+            valueWriter);
+    final var entryLength = recordBatchEntry.getLength();
+
+    if (batchSize + entryLength >= maxBatchSize) {
+      // todo decided whether we want to throw or return a bool or a either?!
```

Comment:
:thought_balloon: I think `Either` is more in keeping with how the engine deals with errors, but as the stream processor is ours, we can decide what we want to do. I personally like `Either` for production code, but it should be whatever our team likes best.

Reply:
I guess Either would fit here